File: gorays/image.go
package main

import (
	"fmt"
	"io"
	"log"
	"os"
)

type image struct {
	size int
	data []byte
}

func newImage(size int) *image {
	return &image{size: size, data: make([]byte, 3*size*size)}
}

func (i *image) Save() {
	f, err := os.Create(*outputfile)
	if err != nil {
		log.Panic(err)
	}
	defer f.Close()

	if err := i.SaveTo(f); err != nil {
		log.Panic(err)
	}
}

func (i *image) SaveTo(w io.Writer) error {
	fmt.Fprintf(w, "P6 %v %v 255 ", i.size, i.size)
	if _, err := w.Write(i.data); err != nil {
		return err
	}
	return nil
}

File: rustrays/image.rs
use std::vec;
use std::io::fs::File;
use std::path::Path;

pub struct Image {
	size: int,
	data: ~[u8]
}

impl Image {
	pub fn new(size: int) -> Image {
		let cap = (3 * size * size) as uint;
		let mut vec = vec::with_capacity(cap);
		unsafe {
			vec.set_len(cap);
		}
		return Image { size: size, data: vec }
	}

	pub fn save(&self, outputfile: &str) {
		let mut file = File::create(&Path::new(outputfile));
		let s = format!("P6 {} {} 255 ", self.size, self.size);
		file.write(s.as_bytes());
		file.write(self.data);
	}
}
